File: src/main/java/kitchenpos/product/domain/model/ProductName.java
package kitchenpos.product.domain.model;

import jakarta.persistence.Embeddable;
import kitchenpos.global.exception.ErrorCode;
import kitchenpos.global.exception.ProfanityException;
import kitchenpos.product.domain.exception.ProductNameException;
import kitchenpos.product.domain.service.ProductPurgomalumClient;

@Embeddable
public record ProductName(String name) {

    public static ProductName of(String name, ProductPurgomalumClient purgomalumClient) {
        if (name == null || name.isBlank()) {
            throw new ProductNameException();
        }

        if (purgomalumClient.containsProfanity(name)) {
            throw new ProfanityException(ErrorCode.PRODUCT_NAME_PROFANITY_NOT_ALLOWED.toString());
        }
        return new ProductName(name);
    }

    public String get() {
        return name;
    }
}

File: src/main/java/kitchenpos/products/domain/ProductName.kt
package kitchenpos.products.domain

class ProductName private constructor(
    val name: String,
) {
    companion object {
        fun create(value: String): ProductName {
            validateNull(value)
            return ProductName(value)
        }

        private fun validateNull(value: String) {
            if (value.isBlank()) {
                throw IllegalArgumentException("Invalid product name: $value")
            }
        }
    }
}
